### PR TITLE
AIP-84 Get Mapped Task Instance Tries

### PR DESCRIPTION
**Pull request by kandharvishnu**
related to: https://github.com/apache/airflow/issues/42370

---

### CODE REVIEW

File: airflow/api_fastapi/core_api/routes/public/task_instances.py [hunk 1]
```diff
@@ -234,6 +236,70 @@ def get_task_instance_dependencies(
...
+    """Get list of task instances history."""
+
+    def _query(orm_object: Base) -> Select:
+        query = select(orm_object).where(
+            orm_object.dag_id == dag_id,
+            orm_object.run_id == dag_run_id,
+            orm_object.task_id == task_id,
+            orm_object.map_index == map_index,
+        )
+        print(type(query))
```
Comment (rawwar) [inline airflow/api_fastapi/core_api/routes/public/task_instances.py]:
```suggestion
```

File: airflow/api_fastapi/core_api/routes/public/task_instances.py [hunk 2]
```diff
@@ -234,6 +236,70 @@ def get_task_instance_dependencies(
...
+@task_instances_router.get(
+    task_instances_prefix + "/{task_id}/tries",
+    responses=create_openapi_http_exception_doc([status.HTTP_404_NOT_FOUND]),
+)
+def get_task_instance_tries(
+    dag_id: str,
+    dag_run_id: str,
+    task_id: str,
+    session: Annotated[Session, Depends(get_session)],
+    map_index: int = -1,
```
Comment (rawwar) [inline airflow/api_fastapi/core_api/routes/public/task_instances.py:250]: map_index will show up in the OpenAPI specifications. So, the user might send map_index here.

File: airflow/api_fastapi/core_api/routes/public/task_instances.py [hunk 3]
```diff
@@ -234,6 +236,70 @@ def get_task_instance_dependencies(
...
+            f"The Task Instance with dag_id: `{dag_id}`, run_id: `{dag_run_id}`, task_id: `{task_id}` and map_index: `{map_index}` was not found",
+        )
+    task_instances = [
+        TaskInstanceHistoryResponse.model_validate(task_instance, from_attributes=True)
+        for task_instance in task_instance_select
+    ]
+    return TaskInstanceHistoryCollectionResponse(
+        task_instances=task_instances,
+        total_entries=len(task_instances),
+    )
```
Comment (rawwar) [inline airflow/api_fastapi/core_api/routes/public/task_instances.py]:
```suggestion
    return TaskInstanceHistoryCollectionResponse(
        task_instances=task_instance_select,
        total_entries=len(task_instance_select),
    )
```

Comment (kandharvishnu) [reply airflow/api_fastapi/core_api/routes/public/task_instances.py]: `TaskInstanceHistoryCollectionResponse` will take the list of `TaskInstanceHistoryResponse`. 

As a first step, I am changing the `airflow.models.taskinstance.TaskInstance` to `TaskInstanceHistoryResponse` and then passing it to `TaskInstanceHistoryCollectionResponse`

Comment (rawwar) [reply airflow/api_fastapi/core_api/routes/public/task_instances.py]: can you use `cast` like here: https://github.com/apache/airflow/blob/2390504d2424aa664a752433c6a868d7b645f72a/airflow/api_fastapi/core_api/routes/public/dag_run.py#L224

File: airflow/api_fastapi/core_api/routes/public/task_instances.py (hunk 2 above)
Comment (ghost) [reply airflow/api_fastapi/core_api/routes/public/task_instances.py:250]: `get_task_instance_tries` will work for non mapped tasks. For mapped tasks which requires map_index, there is a different endpoint `get_mapped_task_instance_tries`.

File: airflow/api_fastapi/core_api/routes/public/task_instances.py (hunk 3 above)
Comment (rawwar) [reply airflow/api_fastapi/core_api/routes/public/task_instances.py]: If this doesn't work. please close this conversation.

**Review by pierrejeambrun — COMMENTED** [review]
Looking good thanks :)

Should be good to merge once the base branch is updated and merged.

I'll do a final review then.

File: airflow/api_fastapi/core_api/routes/public/task_instances.py [hunk 4]
```diff
@@ -469,7 +489,7 @@ def get_task_instance_try_details(
     task_id: str,
     task_try_number: int,
     session: Annotated[Session, Depends(get_session)],
-    map_index: int = -1,
+    map_index: Annotated[int, Depends(lambda: -1)],
```
Comment (pierrejeambrun) [inline airflow/api_fastapi/core_api/routes/public/task_instances.py]: Why this change ? This seems odd, what are we trying to achieve ?

**Review by pierrejeambrun — COMMENTED** [review]
Just one question

File: airflow/api_fastapi/core_api/routes/public/task_instances.py (hunk 4 above)
Comment (kandharvishnu) [reply airflow/api_fastapi/core_api/routes/public/task_instances.py]: changed the default value back to `-1`

**Review by pierrejeambrun — APPROVED** [review]

**Pull request merged**